### ISSUE TITLE
Add ability to attach vispy Filter objects to layer visual

### ISSUE THREAD

**Issue by kevinyamauchi** [napari/napari#3669]
## 🚀 Feature

I think it would be nice if we allowed users to attach vispy shader filters to layer visuals as a way to enable more customized rendering.

Some examples of things one can do with filters filters:
- wireframe for the mesh visual: https://github.com/napari/napari/issues/2970
- clipping planes (implemented by @brisvag ): https://github.com/vispy/vispy/blob/main/examples/scene/clipping_planes.py
- confine rendering to other shapes (e.g., sphere)
- discard fragments that are "background" (i.e., if it is a background value, don't render it so it doesn't cause blending issues with other layers)

cc @maweigert

## Motivation
It is nice to be able to experiment with custom shaders without having to go deep into the codebase. While allowing full shaders to be passed in might be a bit much, the vispy `[BaseFilter](https://vispy.org/api/vispy.visuals.filters.base_filter.html)` might be a nice compromise. 

Also, see the twitter thread below about allowing users to pass in custom shaders.
https://twitter.com/martweig/status/1460643240588238849

## Pitch

We add an evented list to the layer models that accepts vispy `BaseFilter` subclasses. When the list contents change (i.e., a filter is added/removed), an event is emitted and the visual updates accordingly. 

## Alternatives

We could register/coordinate the filters at the viewer level, but it seems like they should be restricted to the layer they are acting on.

**Comment by brisvag** [comment]
Definitely useful! Just for clarity:
- clipping_planes are already hooked up in napari layers with `experimental_clipping_planes`
- mesh filters will be super easy to implement after #3494 is merged (though you can already do it technically)

Clipping based on shapes is gonna be awesome!

**Comment by maweigert** [comment]
Yes, this would be very useful! Ideally, a napari layer would be able to attach a given filter to its visual, which currently needs a slight detour (or is there a better way?) 

```python
layer = ... 
filter = ...

viewer.window.qt_viewer.layer_to_visual[layer].attach(filter)
```

**Comment by brisvag** [comment]
That's the right way, for now!

The only communication between napari layers and vispy visuals is through events, so as kevin suggested we need to add an evented model for filters :)
